components/sup/src/manager/file_watcher.rs

```diff
@@ -1248,2217 +327,1778 @@ impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
+        if l == None {
+            return false;
         }
-        if let Some(ref path) = paths.real_file {
-            actions.push(PathsAction::NotifyFileDisappeared(path.clone()));
+
+        let r = std::fs::canonicalize(right).ok();
+        if r == None {
+            return false;
         }
-        actions.push(PathsAction::ProcessPathAfterSettle(pad.args));
-        actions
+
+        let result = l == r;
+        trace!("FileWatcher::canonical_compare(..)\n  left: {:?}\n right: {:?}\nresult: {:?}",
+               left,
+               right,
+               result);
+        result
     }
 
-    // Maps filesystem events to high-level actions.
-    fn get_event_actions(paths: &Paths, event: DebouncedEvent) -> Vec<EventAction> {
-        // Usual actions on files and resulting events (assuming that
-        // a and b are in the same directory which we watch)
-        // touch a - Create(a)
-        // ln -sf foo a (does not matter if symlink a did exist before)- Create(a)
-        // mkdir a - Create(a)
-        // mv a b (does not matter if b existed or not) - NoticeRemove(a), Rename(a, b)
-        // mv ../a . - Create(a)
-        // mv a .. - NoticeRemove(a), Remove(a)
-        // rm a - NoticeRemove(a), Remove(a)
-        // echo foo >a (assuming a did exist earlier) - NoticeWrite(a), Write(a)
-        let event_action = match event {
-            // `Write` event will handle that.
-            DebouncedEvent::NoticeWrite(_) => EventAction::Ignore,
-            // These happen for regular files, just check if it
-            // affects the file we are watching.
-            //
-            // TODO(krnowak): I wonder if we should watch `Chmod`
-            // events for directories too. Maybe some permission
-            // changes can cause the directory to be unwatchable. Or
-            // watchable again for that matter.
-            DebouncedEvent::Write(ref p) | DebouncedEvent::Chmod(ref p) => {
-                match paths.get_watched_file(p) {
-                    Some(&WatchedFile::Regular(_)) => EventAction::PlainChange(p.clone()),
-                    _ => EventAction::Ignore,
-                }
-            }
-            DebouncedEvent::NoticeRemove(ref p) => Self::handle_notice_remove_event(paths, p),
-            DebouncedEvent::Remove(p) => Self::handle_remove_event(paths, p),
-            DebouncedEvent::Create(ref p) => {
-                match paths.get_watched_file(p) {
-                    None => EventAction::Ignore,
-                    Some(&WatchedFile::MissingRegular(ref c)) => {
-                        EventAction::AddRegular(c.get_paths_action_data())
+    fn handle_file_appearance(&mut self, notify_event: &Event) {
+        trace!("FileWatcher::handle_file_appearance\n| {:?}\n| {:?}\n",
+               notify_event,
+               self.watched_file);
+        for path_buf in notify_event.paths.iter() {
+            let mut should_signal = self.watched_file.path == *path_buf;
```

Comment:
this first check seems unnecessary. Any reason why we would not always do a cannonical compare?

```
if path_buf.is_file() && Self::canonical_compare(&self.watched_file.path, path_buf) {
```

Reply:
I actually had to back some of this out in order to handle differences between the polling and native backends.  The appear/disappear signalling is still combined by the signaling around modification is its own implementation.

Reply:
This hasn't been marked resolved yet, now shows as outdated, and some of your new comments I think circle back to this so will continue the discussion there.